src/core/net/srp_server.cpp

```diff
@@ -0,0 +1,1604 @@
+
+    SuccessOrExit(error = mSocket.SendTo(*response, aMessageInfo));
+
+    if (aResponseCode != Dns::UpdateHeader::kResponseSuccess)
+    {
+        otLogInfoSrp("server: sends fail response: %d", aResponseCode);
+    }
+    else
+    {
+        otLogInfoSrp("server: sends success response");
+    }
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        otLogWarnSrp("server: failed to send response: %s", otThreadErrorToString(error));
+        FreeMessage(response);
+    }
+}
+
+void Server::SendResponse(const Dns::UpdateHeader &aHeader,
+                          uint32_t                 aLease,
+                          uint32_t                 aKeyLease,
+                          const Ip6::MessageInfo & aMessageInfo)
+{
+    otError           error;
+    Message *         response = nullptr;
+    Dns::UpdateHeader header;
+    const char        kRootName[2] = ".";
+    Dns::OptRecord    optRecord;
+    Dns::LeaseOption  leaseOption;
+
+    response = mSocket.NewMessage(0);
+    VerifyOrExit(response != nullptr, error = OT_ERROR_NO_BUFS);
+
+    header.SetMessageId(aHeader.GetMessageId());
+    header.SetType(Dns::UpdateHeader::kTypeResponse);
+    header.SetQueryType(aHeader.GetQueryType());
+    header.SetResponseCode(Dns::UpdateHeader::kResponseSuccess);
+    header.SetAdditionalRecordCount(1);
+    SuccessOrExit(error = response->Append(header));
+
+    SuccessOrExit(error = Dns::Name::AppendName(kRootName, *response));
+
+    optRecord.Init();
+    optRecord.SetUdpPayloadSize(kUdpPayloadSize);
+    optRecord.SetDnsSecurityFlag();
+    optRecord.SetLength(sizeof(Dns::LeaseOption));
+    SuccessOrExit(error = response->Append(optRecord));
+
+    leaseOption.Init();
+    leaseOption.SetLeaseInterval(aLease);
+    leaseOption.SetKeyLeaseInterval(aKeyLease);
+    SuccessOrExit(error = response->Append(leaseOption));
+
+    SuccessOrExit(error = mSocket.SendTo(*response, aMessageInfo));
+
+    otLogInfoSrp("server: sent response with granted lease: %u and key lease: %u", aLease, aKeyLease);
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        otLogWarnSrp("server: failed to send response: %s", otThreadErrorToString(error));
+        FreeMessage(response);
+    }
+}
+
+void Server::HandleUdpReceive(void *aContext, otMessage *aMessage, const otMessageInfo *aMessageInfo)
+{
+    static_cast<Server *>(aContext)->HandleUdpReceive(*static_cast<Message *>(aMessage),
+                                                      *static_cast<const Ip6::MessageInfo *>(aMessageInfo));
+}
+
+void Server::HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    otError           error;
+    Dns::UpdateHeader dnsHeader;
+    uint16_t          offset = aMessage.GetOffset();
+
+    SuccessOrExit(error = aMessage.Read(offset, dnsHeader));
+    offset += sizeof(dnsHeader);
+
+    // Handles only queries.
+    VerifyOrExit(dnsHeader.GetType() == Dns::UpdateHeader::Type::kTypeQuery, error = OT_ERROR_DROP);
```

Comment:
```suggestion
    VerifyOrExit(dnsHeader.GetType() == Dns::UpdateHeader::Type::kTypeQuery, error = OT_ERROR_NOT_IMPLEMENTED);
```

should respond with DNS RCODE NOT_IMPLEMENTED ?

Reply:
I think the SRP server doesn't need to handle standard queries (which should be handled by a DNS-SD server), so `OT_ERROR_DROP` I think is better here.